### PR TITLE
[MCLEAN-93] Support junctions on NTFS

### PR DESCRIPTION
**Pull request by gnodet**
Jira issue: [MCLEAN-93](https://issues.apache.org/jira/projects/MCLEAN/issues/MCLEAN-93)

### CODE REVIEW

**Review by slachiewicz — APPROVED** [review]

**Comment by gnodet** [comment]
I'm investigating if I can setup a test on windows.  I've been sidetracked by trying to switch the tests to junit 5 / plexus-testing, but it's missing the helper methods to setup/lookup mojos...
I'll provide one asap.

File: src/main/java/org/apache/maven/plugins/clean/Cleaner.java [hunk 1]
```diff
@@ -297,6 +299,12 @@ else if ( file.exists() )
         return result;
     }
 
+    private boolean isSymbolicLink( Path path ) throws IOException
+    {
+        BasicFileAttributes attrs = Files.readAttributes( path, BasicFileAttributes.class, LinkOption.NOFOLLOW_LINKS );
+        return attrs.isSymbolicLink() || ( attrs.isDirectory() && attrs.isOther() );
```
Comment (michael-o) [inline src/main/java/org/apache/maven/plugins/clean/Cleaner.java:305]: Can you add a comment here  that the right hand part represents a junction?

**Review by michael-o — COMMENTED** [review]
When can this happen? I mean, can we create a junction in Java without using native code?

File: src/main/java/org/apache/maven/plugins/clean/Cleaner.java (hunk 1 above)
Comment (gnodet) [reply src/main/java/org/apache/maven/plugins/clean/Cleaner.java:305]: I'm experimenting with something like:
```
        Path testDir = Paths.get( "target/test-dir" ).toAbsolutePath();
        Files.createDirectories( testDir );
        Path orgDir = testDir.resolve( "org-dir" );
        Path jctDir = testDir.resolve( "jct-dir" );
        Files.createDirectories( orgDir );

        Process process = new ProcessBuilder()
                .directory( testDir.toFile() )
                .command( "mklink", "/J", jctDir.toString(), orgDir.toString() )
                .start();
        int result = process.waitFor();
```
The idea comes from https://www.geeksforgeeks.org/creating-junction-points/

**Comment by gnodet** [comment]
> When can this happen? I mean, can we create a junction in Java without using native code?

I'm experimenting with something like:
```
        Path testDir = Paths.get( "target/test-dir" ).toAbsolutePath();
        Files.createDirectories( testDir );
        Path orgDir = testDir.resolve( "org-dir" );
        Path jctDir = testDir.resolve( "jct-dir" );
        Files.createDirectories( orgDir );

        Process process = new ProcessBuilder()
                .directory( testDir.toFile() )
                .command( "mklink", "/J", jctDir.toString(), orgDir.toString() )
                .start();
        int result = process.waitFor();
```
The idea comes from https://www.geeksforgeeks.org/creating-junction-points/

**Comment by gnodet** [comment]
> When can this happen? I mean, can we create a junction in Java without using native code?

The test is here: https://github.com/apache/maven-clean-plugin/pull/10/files#diff-f51e0a6f6bcc199bb061d2f1b751c0740ae8cd016a1e87b5349758c4ee96e7d1R297-R321

**Comment by jvoccia** [comment]
> When can this happen? I mean, can we create a junction in Java without using native code?

Not sure if you are asking about this regarding "tests" or regarding what "real world" use case this could occur in, but if asking in the second case:

NPM does this on Windows.  So if you link a NPM component with a package.json like:
```json
{
  ...
  "dependencies": {
    "someTestServer": "file:src/test/node_server/someTestServer"
  }
}
```
  
  You will get a node_modules directory like:
  node_modules/someTestServer <- Junction Point pointing to: src/test/node_server/someTestServer
  
 If you configure mvn clean to delete node_modules (or a directory above node_modules) - it will walk the junction point and delete your src files.

**Comment by jira-importer** [comment]
Resolve #174

**Pull request merged**